mypy/meet.py

```diff
@@ -221,9 +221,32 @@ def is_none_typevar_overlap(t1: Type, t2: Type) -> bool:
     # As before, we degrade into 'Instance' whenever possible.
 
     if isinstance(left, TypeType) and isinstance(right, TypeType):
-        # TODO: Can Callable[[...], T] and Type[T] be partially overlapping?
         return _is_overlapping_types(left.item, right.item)
 
+    # Type[C] vs Callable[..., C], where the latter is class object.
+    if isinstance(left, CallableType) and left.is_type_obj() and isinstance(right, TypeType):
+        if _is_overlapping_types(left.ret_type, right.item):
+            return True
+    if isinstance(right, CallableType) and right.is_type_obj() and isinstance(left, TypeType):
```

Comment:
I'd say define a nested function for doing the `Type[C]` vs. whatever checks instead of duplicating the logic for the left and right cases.